File: laurus-python/examples/hybrid_search.py
"""Hybrid Search Example — combining lexical and vector search.

Demonstrates:
- Lexical-only search (for comparison)
- Vector-only search (for comparison)
- Hybrid search with RRF fusion
- Hybrid search with WeightedSum fusion
- Hybrid search with a filter query

The embedder is registered in the schema and laurus automatically converts
text to vectors at index and query time — no external embedding library needed.

Run with:
    maturin develop --features embeddings-candle
    python examples/hybrid_search.py
"""

from __future__ import annotations

import laurus

# ---------------------------------------------------------------------------
# Embedder configuration
# ---------------------------------------------------------------------------

_EMBEDDER_NAME = "bert"
_EMBEDDER_MODEL = "sentence-transformers/all-MiniLM-L6-v2"
_DIM = 384  # dimension for all-MiniLM-L6-v2

# ---------------------------------------------------------------------------
# Dataset
# ---------------------------------------------------------------------------

CHUNKS = [
    ("django_guide", "Django Web Development", "Django follows the model-template-view architecture pattern for clean separation of concerns.", 1, "framework"),
    ("django_guide", "Django Web Development", "Django ORM maps Python classes to database tables with migrations for schema management.", 2, "framework"),
    ("django_guide", "Django Web Development", "Django middleware processes requests and responses through a chain of pluggable components.", 3, "framework"),
    ("flask_guide", "Flask Microservices", "Flask provides lightweight routing and Jinja2 templating for building APIs and web apps.", 1, "framework"),
    ("flask_guide", "Flask Microservices", "Flask extensions like Flask-SQLAlchemy and Flask-Migrate add database support to Flask projects.", 2, "framework"),
    ("numpy_docs", "NumPy Fundamentals", "NumPy arrays provide vectorized operations that are much faster than Python loops for numerical computing.", 1, "scientific"),
    ("numpy_docs", "NumPy Fundamentals", "Broadcasting in NumPy allows arithmetic operations on arrays of different shapes without copying data.", 2, "scientific"),
    ("pytest_book", "Testing with pytest", "pytest fixtures provide reusable setup and teardown logic for test functions with dependency injection.", 1, "testing"),
    ("pytest_book", "Testing with pytest", "pytest parametrize decorator runs the same test with different input datasets automatically.", 2, "testing"),
]


def main() -> None:
    print("=== Laurus Hybrid Search Example ===\n")
    print(f"Embedder: {_EMBEDDER_MODEL} (dim={_DIM})\n")

    # ── Schema ─────────────────────────────────────────────────────────────
    schema = laurus.Schema()
    schema.add_embedder(_EMBEDDER_NAME, {"type": "candle_bert", "model": _EMBEDDER_MODEL})
    schema.add_text_field("title")
    schema.add_text_field("text")
    schema.add_text_field("category")
    schema.add_integer_field("page")
    schema.add_flat_field("text_vec", dimension=_DIM, distance="cosine", embedder=_EMBEDDER_NAME)
    schema.set_default_fields(["text"])

    index = laurus.Index(schema=schema)

    # ── Index ──────────────────────────────────────────────────────────────
    # Passing raw text to text_vec lets laurus automatically vectorize it via the built-in embedder.
    print("--- Indexing chunked documents ---\n")
    for doc_id, title, text, page, category in CHUNKS:
        index.add_document(
            doc_id,
            {
                "title": title,
                "text": text,
                "category": category,
                "page": page,
                "text_vec": text,
            },
        )
    index.commit()
    print(f"Indexed {len(CHUNKS)} chunks.\n")

    # =====================================================================
    # [A] Lexical-only search (baseline)
    # =====================================================================
    print("=" * 60)
    print("[A] Lexical-only: term 'fixtures' in text")
    print("=" * 60)
    _print_results(index.search(laurus.TermQuery("text", "fixtures"), limit=3))

    # =====================================================================
    # [B] Vector-only search (baseline)
    # =====================================================================
    print("\n" + "=" * 60)
    print("[B] Vector-only: 'database query builder'")
    print("=" * 60)
    _print_results(
        index.search(laurus.VectorTextQuery("text_vec", "database query builder"), limit=3)
    )

    # =====================================================================
    # [C] Hybrid search — RRF Fusion (default when both queries present)
    # =====================================================================
    print("\n" + "=" * 60)
    print("[C] Hybrid (RRF k=60): vector='template rendering' + lexical='jinja2'")
    print("=" * 60)
    request = laurus.SearchRequest(
        lexical_query=laurus.TermQuery("text", "jinja2"),
        vector_query=laurus.VectorTextQuery("text_vec", "template rendering"),
        fusion=laurus.RRF(k=60.0),
        limit=3,
    )
    _print_results(index.search(request))

    # =====================================================================
    # [D] Hybrid search — WeightedSum Fusion
    # =====================================================================
    print("\n" + "=" * 60)
    print("[D] Hybrid (WeightedSum 0.3/0.7): vector='data analysis' + lexical='arrays'")
    print("=" * 60)
    request = laurus.SearchRequest(
        lexical_query=laurus.TermQuery("text", "arrays"),
        vector_query=laurus.VectorTextQuery("text_vec", "data analysis"),
        fusion=laurus.WeightedSum(lexical_weight=0.3, vector_weight=0.7),
        limit=3,
    )
    _print_results(index.search(request))

    # =====================================================================
    # [E] Hybrid search with filter query
    # =====================================================================
    print("\n" + "=" * 60)
    print("[E] Hybrid + filter: vector='testing automation' + lexical='parametrize' + category='testing'")
    print("=" * 60)
    request = laurus.SearchRequest(
        lexical_query=laurus.TermQuery("text", "parametrize"),
        vector_query=laurus.VectorTextQuery("text_vec", "testing automation"),
        filter_query=laurus.TermQuery("category", "testing"),
        fusion=laurus.RRF(k=60.0),
        limit=3,
    )
    _print_results(index.search(request))

    # =====================================================================
    # [F] Hybrid search via SearchRequest with DSL string query
    # =====================================================================
    print("\n" + "=" * 60)
    print("[F] DSL query string: 'text:middleware'")
    print("=" * 60)
    _print_results(index.search("text:middleware", limit=3))

    print("\nHybrid search example completed!")


def _print_results(results: list) -> None:
    if not results:
        print("  (no results)")
        return
    for r in results:
        doc = r.document or {}
        text = doc.get("text", "")
        print(f"  id={r.id!r:8s}  score={r.score:.4f}  text={text!r:.60s}")


if __name__ == "__main__":
    main()

File: laurus-php/examples/hybrid_search.php
<?php

// Hybrid Search Example — combining lexical and vector search.
//
// Demonstrates:
// - Lexical-only search (for comparison)
// - Vector-only search (for comparison)
// - Hybrid search with RRF fusion
// - Hybrid search with WeightedSum fusion
// - Hybrid search with a filter query
//
// The embedder is registered in the schema and laurus automatically converts
// text to vectors at index and query time — no external embedding library needed.
//
// Usage:
//   cd laurus-php
//   cargo build --release --features embeddings-candle
//   php -d extension=target/release/liblaurus_php.so examples/hybrid_search.php

use Laurus\Index;
use Laurus\RRF;
use Laurus\Schema;
use Laurus\SearchRequest;
use Laurus\TermQuery;
use Laurus\VectorTextQuery;
use Laurus\WeightedSum;

const EMBEDDER_NAME = "bert";
const EMBEDDER_MODEL = "sentence-transformers/all-MiniLM-L6-v2";
const DIM = 384;

$chunks = [
    ["laravel", "Laravel Framework", "Laravel provides an elegant MVC architecture with Eloquent ORM for database operations.", 1, "framework"],
    ["laravel", "Laravel Framework", "Blade is Laravel's templating engine that compiles to plain PHP for maximum performance.", 2, "framework"],
    ["laravel", "Laravel Framework", "Laravel Artisan CLI provides commands for migrations, seeding, and code generation.", 3, "framework"],
    ["symfony", "Symfony Components", "Symfony's dependency injection container manages service instantiation and configuration.", 1, "framework"],
    ["symfony", "Symfony Components", "Symfony's HttpFoundation component provides an object-oriented layer for HTTP requests and responses.", 2, "framework"],
    ["wordpress", "WordPress Internals", "WordPress uses hooks (actions and filters) to allow plugins to modify core behavior without editing source.", 1, "cms"],
    ["wordpress", "WordPress Internals", "The WordPress REST API enables headless CMS architectures with JSON endpoints.", 2, "cms"],
    ["phpunit", "PHPUnit Testing", "PHPUnit data providers allow running the same test with different input datasets automatically.", 1, "testing"],
    ["phpunit", "PHPUnit Testing", "Mock objects in PHPUnit simulate dependencies to isolate the unit under test.", 2, "testing"],
];

function print_results(array $results): void
{
    if (empty($results)) {
        echo "  (no results)" . PHP_EOL;
        return;
    }
    foreach ($results as $r) {
        $doc = $r->getDocument();
        $text = substr($doc["text"] ?? "", 0, 70);
        printf("  id=%-12s  score=%.4f  text='%s'\n", "'{$r->getId()}'", $r->getScore(), $text);
    }
}

echo "=== Laurus Hybrid Search Example ===" . PHP_EOL . PHP_EOL;
echo "Embedder: " . EMBEDDER_MODEL . " (dim=" . DIM . ")" . PHP_EOL . PHP_EOL;

// ── Schema ─────────────────────────────────────────────────────────────
$schema = new Schema();
$schema->addEmbedder(EMBEDDER_NAME, ["type" => "candle_bert", "model" => EMBEDDER_MODEL]);
$schema->addTextField("title");
$schema->addTextField("text");
$schema->addTextField("category");
$schema->addIntegerField("page");
$schema->addFlatField("text_vec", DIM, "cosine", EMBEDDER_NAME);
$schema->setDefaultFields(["text"]);

$index = new Index(null, $schema);

// ── Index ──────────────────────────────────────────────────────────────
echo "--- Indexing chunked documents ---" . PHP_EOL . PHP_EOL;
foreach ($chunks as [$docId, $title, $text, $page, $category]) {
    $index->addDocument($docId, [
        "title" => $title,
        "text" => $text,
        "category" => $category,
        "page" => $page,
        "text_vec" => $text,
    ]);
}
$index->commit();
echo "Indexed " . count($chunks) . " chunks." . PHP_EOL . PHP_EOL;

// =====================================================================
// [A] Lexical-only search (baseline)
// =====================================================================
echo str_repeat("=", 60) . PHP_EOL;
echo "[A] Lexical-only: term 'eloquent' in text" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;
print_results($index->search(new TermQuery("text", "eloquent"), 3));

// =====================================================================
// [B] Vector-only search (baseline)
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "[B] Vector-only: 'database query builder'" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;
print_results($index->search(new VectorTextQuery("text_vec", "database query builder"), 3));

// =====================================================================
// [C] Hybrid search — RRF Fusion
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "[C] Hybrid (RRF k=60): vector='template rendering' + lexical='blade'" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;
$request = new SearchRequest(
    null,
    new TermQuery("text", "blade"),
    new VectorTextQuery("text_vec", "template rendering"),
    null,
    new RRF(60.0),
    3,
);
print_results($index->search($request));

// =====================================================================
// [D] Hybrid search — WeightedSum Fusion
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "[D] Hybrid (WeightedSum 0.3/0.7): vector='API development' + lexical='rest'" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;
$request = new SearchRequest(
    null,
    new TermQuery("text", "rest"),
    new VectorTextQuery("text_vec", "API development"),
    null,
    new WeightedSum(0.3, 0.7),
    3,
);
print_results($index->search($request));

// =====================================================================
// [E] Hybrid search with filter query
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "[E] Hybrid + filter: vector='dependency management' + lexical='injection' + category='framework'" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;
$request = new SearchRequest(
    null,
    new TermQuery("text", "injection"),
    new VectorTextQuery("text_vec", "dependency management"),
    new TermQuery("category", "framework"),
    new RRF(60.0),
    3,
);
print_results($index->search($request));

// =====================================================================
// [F] DSL string query
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "[F] DSL query string: 'text:hooks'" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;
print_results($index->search("text:hooks", 3));

echo PHP_EOL . "Hybrid search example completed!" . PHP_EOL;
